Improve rendering of turing machines

modderme123 opened the pull request:
I wanted a view that I could scroll through the pixels without them being blurry, and see the state that the head was in. Maybe this could be useful for others?

tcosmo commented:
@modderme123 Thank you very much for your contribution! You are officially the first person to contribute to the project and that makes me very happy :) You even got ahead of myself as I still haven't had time to write https://bbchallenge.org/contribute or to populate the forum with categories and posts.

I agree with you that the current visualization is far from optimal and needs improvements: its barely enough just to get the project to take off. Using image rescaling for these traces as it is done now is a bad idea as it blurs details but it was the quickest thing to implement.

I like the idea of your fix as the scroll inherently allows for better resolution. However, I would make it optional (as one of the simulation parameter). Indeed, I think this will vary for each user but for me I personally prefer to just have a rapid, one glance, overview of the machine's behavior at the cost of blurring the details. 

However, I'm none too keen on the color convention for states because:
1- it wont scale easily to machines with more states such as [mAQACAAEEAQADAAAGAQEDAQEBAAEFAAAAAQEBAAACAAADAAAF](https://bbchallenge.org/mAQACAAEEAQADAAAGAQEDAQEBAAEFAAAAAQEBAAACAAADAAAF&s=20000&w=500&ox=0.3&status=halt ) which actually breaks your code
2- I dont believe that this way of integreting state information to the traces is very informative (we barely see the state color in big traces)
3- It complexifies the visual identity of the website

Again, my philosophy with this visualisation tool is primarly to give a condensed picture of a machine's behavior. I'm completely open to have more options to focus on details of the diagrams but the default settings should give the "simplest" image as possible.

Thank you very much for your work!

modderme123 commented:
Those points make sense! We used this simulation in a class (hi @EricESeverson) to analyze the behavior of the current bb5 champion, and wanted to be able to see how the different states interacted. 

I'm happy to break up/redo these changes in whatever way you would like (feel free to close this PR and ask for those changes in a separate one if that would be helpful).

Also, I've now updated the PR so that it should just display white for states that it doesn't know about (and added the initial state option that we used in class).

tcosmo commented:
That's really cool. Many thanks to Eric Severson for sharing bbchallenge.

The only request that I would have would be to make the view be optional via a checkbox "display scrollable view" for instance. Same for the colors, another checkbox "display states with colors" or something better worded.

I have added you as a contributor to the project and created a new PR for this :) https://github.com/bbchallenge/website-frontend/pull/3
